### PR TITLE
refactor: add otel tracing filter to logging

### LINKED ISSUE #458: Create a custom `logging.Filter` that automatically extracts the current OpenTelemetry trace context and injects `trace_id` and `span_id` into every log record
**Description:**
Create a custom `logging.Filter` that automatically extracts the current OpenTelemetry trace context and injects `trace_id` and `span_id` into every log record. This enables correlation between logs and distributed traces.

**Detailed Requirements:**
1. Create `OtelTraceFilter` class (extends `logging.Filter`) that:
   - Gets current span from `opentelemetry.trace.get_current_span()`
   - Extracts `trace_id` and `span_id` from span context
   - Adds fields to `LogRecord` as attributes
   - Handles case when no active span (sets to None or empty string)
2. Add filter to FancyLogger's handlers
3. Update `JsonFormatter` to include trace_id/span_id fields
4. Update `CustomFormatter` to optionally show trace context
5. Ensure zero overhead when OpenTelemetry not installed (graceful degradation)
6. Format trace_id/span_id as hex strings (standard OTEL format)

**Files to Modify:**
- `mellea/core/utils.py` - Add OtelTraceFilter, update formatters

**Acceptance Criteria:**
- [ ] trace_id and span_id appear in all JSON log records when tracing is active
- [ ] Logs work correctly when tracing is disabled (fields are null/absent)
- [ ] No import errors when opentelemetry not installed
- [ ] Filter added to FancyLogger initialization

### PR DESCRIPTION
**Pull request by AngeloDanducci**
# Misc PR
Will need to update after #457 is merged to use MelleaLogger

## Type of PR

- [ ] Bug Fix
- [ ] New Feature
- [ ] Documentation
- [x] Other

## Description
- [x] Link to Issue: Fixes #458 

### Testing
- [x] Tests added to the respective file if code was changed
- [x] New code has 100% coverage if code as added
- [ ] Ensure existing tests and github automation passes (a maintainer will kick off the github automation when the rest of the PR is populated)

### CODE REVIEW

File: mellea/core/utils.py [hunk 1]
```diff
@@ -323,6 +360,11 @@ def _build_log_dict(self, record: logging.LogRecord) -> dict[str, Any]:
         else:
             log_record = {k: v for k, v in all_core.items() if k not in self._exclude}
 
+        # Add trace context if available
+        if hasattr(record, "trace_id"):
+            log_record["trace_id"] = record.trace_id  # type: ignore[attr-defined]
+            log_record["span_id"] = record.span_id  # type: ignore[attr-defined]
```
Comment (ajbozarth) [inline mellea/core/utils.py]: `span_id` is accessed unconditionally whenever `trace_id` is present, but there is no guard against it being absent. Any code that sets only `trace_id` on a record (e.g. a custom filter, or `set_log_context(trace_id="x")` once the reservation gap is fixed) will raise `AttributeError` at log time.

`CustomFormatter.format()` in this same PR uses the safer pattern — suggest aligning:

```python
if hasattr(record, "trace_id"):
    log_record["trace_id"] = record.trace_id  # type: ignore[attr-defined]
    span_id = getattr(record, "span_id", None)
    if span_id is not None:
        log_record["span_id"] = span_id
```

File: test/core/test_utils_logging.py [hunk 2]
```diff
@@ -114,10 +223,10 @@ def teardown_method(self) -> None:
         clear_log_context()
 
     def test_context_fields_appear_in_output(self) -> None:
-        set_log_context(trace_id="abc-123")
+        set_log_context(request_id="abc-123")
```
Comment (ajbozarth) [inline test/core/test_utils_logging.py:238]: `trace_id` and `span_id` are not in `RESERVED_LOG_RECORD_ATTRS`, so `set_log_context(trace_id="my-id")` passes validation silently. `ContextFilter` stamps the record, then `OtelTraceFilter` silently overwrites it when a span is active.

The test rename from `trace_id="abc-123"` → `request_id="abc-123"` signals awareness of the conflict, but the validation guard was not updated. Suggest adding both names to the frozenset so `set_log_context` raises `ValueError` for them, consistent with the other reserved attrs.

**Review by ajbozarth — COMMENTED** [review]
Edit: I tried letting Claude review and post this time while I was working on a different task

## Review: OTel Trace Context Injection

Clean, well-motivated addition. The no-op guard when OTel isn't installed is correct, the test coverage is thorough, and the `_otel_span` context manager is a solid test helper. A few issues worth addressing before merging.

**Bugs:**
1. Unsafe `span_id` access in `JsonFormatter._build_log_dict` — crashes if `span_id` is absent when `trace_id` is present
2. `trace_id` / `span_id` not added to `RESERVED_LOG_RECORD_ATTRS` — silent overwrite risk from `set_log_context`

**Design / Clarity:**
3. Integration tests mutate the `MelleaLogger` singleton without resetting it first — order-dependent flakiness risk

**Nits:**
- Blank line removal before `# Apply include/exclude filtering` is an unrelated cosmetic change
- Docstring update on private `_build_log_dict` is redundant; the public `format()` already got the note

File: test/core/test_utils_logging.py [hunk 3]
```diff
@@ -507,3 +616,140 @@ def test_all_default_fields_accepted(self) -> None:
...
+            result = fmt.format(record)
+            assert isinstance(result, str)
+
+
+# ---------------------------------------------------------------------------
+# Integration: filter → formatter round-trip
+# ---------------------------------------------------------------------------
+
+
+class TestFilterFormatterIntegration:
```
Comment (ajbozarth) [inline test/core/test_utils_logging.py:705]: `TestFilterFormatterIntegration` mutates the `MelleaLogger` singleton without resetting it before each test. If the singleton was already initialized by a prior test class, `OtelTraceFilter` may or may not be attached depending on execution order.

`TestMelleaLoggerFiltersWired` resets the singleton in `setup_method` — suggest doing the same here:

```python
def setup_method(self) -> None:
    MelleaLogger.logger = None
    logging.getLogger("fancy_logger").handlers.clear()
    logging.getLogger("fancy_logger").filters.clear()

def teardown_method(self) -> None:
    MelleaLogger.logger = None
    logging.getLogger("fancy_logger").handlers.clear()
    logging.getLogger("fancy_logger").filters.clear()
```

File: test/core/test_utils_logging.py (hunk 2 above)
Comment (AngeloDanducci) [reply test/core/test_utils_logging.py:238]: `trace_id` and `span_id` added to `RESERVED_LOG_RECORD_ATTRS`

File: mellea/core/utils.py [hunk 4]
```diff
@@ -323,6 +362,13 @@ def _build_log_dict(self, record: logging.LogRecord) -> dict[str, Any]:
         else:
             log_record = {k: v for k, v in all_core.items() if k not in self._exclude}
 
+        # Add trace context if available
+        if hasattr(record, "trace_id"):
+            log_record["trace_id"] = record.trace_id  # type: ignore[attr-defined]
+            span_id = getattr(record, "span_id", None)
+            if span_id is not None:
+                log_record["span_id"] = span_id
```
Comment (jakelorocco) [inline mellea/core/utils.py:370]: Is it possible to have only a trace_id or only a span_id?

File: test/core/test_utils_logging.py [hunk 5]
```diff
@@ -114,33 +235,33 @@ def teardown_method(self) -> None:
...
-        set_log_context(trace_id="abc-123")
+        set_log_context(request_id="abc-123")
         fmt = JsonFormatter()
         parsed = json.loads(fmt.format(_make_record()))
-        assert parsed.get("trace_id") == "abc-123"
+        assert parsed.get("request_id") == "abc-123"
 
     def test_multiple_context_fields(self) -> None:
-        set_log_context(trace_id="t1", request_id="r1", user="alice")
+        set_log_context(custom_trace="t1", request_id="r1", user="alice")
```
Comment (jakelorocco) [inline test/core/test_utils_logging.py:244]: To clarify, these fields for "custom_trace", "request_id", etc... could be anything for these tests? They just previously happened to coincide with what our otel keys ended up being named?

**Review by jakelorocco — APPROVED** [review]
A few questions but lgtm without more changes. Thank you

File: mellea/core/utils.py (hunk 4 above)
Comment (AngeloDanducci) [reply mellea/core/utils.py:370]: The change in utils.py will always assign both.

The check for None here catches if this is somehow modified without using the filter directly via OtelTraceFilter.  In the reverse case (`span_id` set but `trace_id` not set) the check on line 366 will skip over it entirely and neither will be emitted.

So technically yes but in practice you'd never see the output and the check her catches the edge the alternative edge case.

File: test/core/test_utils_logging.py (hunk 5 above)
Comment (AngeloDanducci) [reply test/core/test_utils_logging.py:244]: Correct they can be anything - and were fine as is before.  In the most recent commit though I added `trace_id` and `span_id` to the `RESERVED_LOG_RECORD_ATTRS` to prevent them from being silently overwritten via `set_log_context` so the naming here had to change.

**Comment by AngeloDanducci** [comment]
I believe these are already addressed @ajbozarth 

1. conditional span_id ref https://github.com/generative-computing/mellea/blob/main/mellea/core/utils.py#L368
2. reserved log record attrs - https://github.com/generative-computing/mellea/blob/main/mellea/core/utils.py#L70, https://github.com/generative-computing/mellea/blob/main/mellea/core/utils.py#L74
3.  Setup and teardown here https://github.com/generative-computing/mellea/blob/main/test/core/test_utils_logging.py#L706

**Comment by ajbozarth** [comment]
> I believe these are already addressed @ajbozarth

Yeah I deleted my comment, I didn't noticed that since the branch had been deleted already that I hadn't actually pulled latest code for Claude to look at
